src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturation.cpp

```diff
@@ -0,0 +1,129 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file ControlAllocationSequentialDesaturation.cpp
+ *
+ * @author Roman Bapst <bapstroman@gmail.com>
+ */
+
+#include "ControlAllocationSequentialDesaturation.hpp"
+
+
+
+void
+ControlAllocationSequentialDesaturation::allocate()
+{
+	// Compute new gains if needed
+	if (_A_update_needed) {
+		_A = matrix::geninv(_B);
+		_A_update_needed = false;
+	}
+
+	// Allocate
+	_actuator_sp = _A * _control_sp;
+
+	// go through control axes from lowest to highest priority and unsaturate the actuators
+	for (unsigned i = 0; i < NUM_AXES; i++) {
+		_actuator_sp = desaturateActuators(_actuator_sp, _axis_prio_increasing[i]);
+	}
+
+	// Clip
+	_actuator_sp = clipActuatorSetpoint();
+
+	// Compute achieved control
+	_control_allocated = _B * _actuator_sp;
+}
+
+ControlAllocation::ActuatorVector ControlAllocationSequentialDesaturation::desaturateActuators(
+	ActuatorVector actuator_sp,
+	ControlAxis axis)
+{
+	ActuatorVector desaturation_vector = getDesaturationVector(axis);
+
+	float gain = computeDesaturationGain(desaturation_vector, actuator_sp);
+
+	actuator_sp = actuator_sp + gain * desaturation_vector;
+
+	gain = computeDesaturationGain(desaturation_vector, actuator_sp);
+
+	actuator_sp = actuator_sp + 0.5f * gain * desaturation_vector;
+
+	return actuator_sp;
+
+}
+
+ControlAllocation::ActuatorVector ControlAllocationSequentialDesaturation::getDesaturationVector(ControlAxis axis)
+{
+	ActuatorVector ret;
+
+	for (unsigned i = 0; i < NUM_ACTUATORS; i++) {
+		ret(i) = _A(i, axis);
```

Comment:
Micro-optimisation: you may use matrix::slice and return a const reference

Reply:
@jlecoeur Can you have a look if it's correct what I did? I was having trouble returning a reference, compiler was complaining about it (Returning reference to temporary object).